File: week3/java_week3/Pen&Paper/Paper.java
public class Paper {
    private int maxSymbols;
    private int symbols;
    private StringBuilder content;

    public Paper(int maxSymbols) {
        this.maxSymbols = maxSymbols;
        this.symbols = 0;
        this.content = new StringBuilder();
    }

    public Paper() {
        this(4096);
    }

    public int getMaxSymbols() {
        return maxSymbols;
    }

    public int getSymbols() {
        return symbols;
    }

    public void addContent(String message) throws OutOfSpaceException {
        int symbolsLeft;

        if ( maxSymbols < 0 ) {
            System.out.println("Wrong max symbol param.");
            return;
        } 

        if ( symbols == maxSymbols ) {
            throw new OutOfSpaceException();
        }

        symbolsLeft = maxSymbols - symbols;

        if ( message.length() > symbolsLeft ) {
            content.append(message.substring(0, symbolsLeft));
            symbols = maxSymbols;
            
            return;
        }

        content.append(message);
        symbols += message.length();
    }

    public void show() {
        if ( symbols == 0 ) {
            System.out.println("Paper is empty.");
            return;
        }

        System.out.println(content);
    }
}

File: week3/cpp_hw_week3/Pen&Paper/Paper.cpp
#include <string>
#include "Paper.h"

Paper::Paper(int maxSymbols) {
    this->maxSymbols = maxSymbols;
    this->symbols = 0;
}

Paper::~Paper() {

}

int Paper::getMaxSymbols() const {
    return this->maxSymbols;
}

int Paper::getSymbols() const {
    return this->symbols;
}

void Paper::addContent(const std::string& message) {
    int symbolsLeft;

    if ( this->maxSymbols < 0 ) {
        std::cout << "Wrong max symbol param." << std::endl;
        return;
    }
    
    if ( this->symbols == this->maxSymbols ) {
        throw OutOfSpace();
    }

    symbolsLeft = this->maxSymbols - this->symbols;

    if ( message.length() > symbolsLeft ) {
        this->content +=  message.substr(0, symbolsLeft);
        this->symbols = this->maxSymbols;
        return;;
    }
    
    this->content += message;
    this->symbols += message.length();
}

void Paper::show() const {
    if ( this->symbols == 0 ) {
        std::cout << "Paper is empty." << std::endl;
        return;
    }

    std::cout << this->content << std::endl;
}